<!DOCTYPE html><html lang="en"><head><title>Static News</title><meta charSet="utf-8"/><meta name="description" content="Static delayed Hacker News."/><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"/><meta name="theme-color" media="(prefers-color-scheme: dark)" content="#1d1f21"/><meta name="viewport" content="width=device-width,initial-scale=1.0"/><meta name="application-name" content="Static News"/><meta name="apple-mobile-web-app-title" content="Static News"/><meta name="apple-mobile-web-app-capable" content="yes"/><meta name="mobile-web-app-capable" content="yes"/><meta name="apple-mobile-web-app-status-bar-style" content="#1d1f21"/><link rel="preload" href="styles.css?v=1719738059202" as="style"/><link rel="stylesheet" href="styles.css?v=1719738059202"/></head><body><div id="container"><div id="inner"><header><a href="/">Static News</a><a href="/about">about</a></header><div id="content"><div><div id="title"><a href="https://adamhooper.medium.com/my-philosophy-of-exceptions-theyre-always-ambiguous-9b419d2ccbef">My philosophy of exceptions: they&#x27;re always ambiguous (2021)</a> <span class="domain">(<a href="https://adamhooper.medium.com">adamhooper.medium.com</a>)</span></div><div class="subtext"><span>zbentley</span> | <span>25 comments</span></div><br/><div><div id="40834905" class="c"><input type="checkbox" id="c-40834905" checked=""/><div class="controls bullet"><span class="by">wvenable</span><span>|</span><a href="#40835763">next</a><span>|</span><label class="collapse" for="c-40834905">[-]</label><label class="expand" for="c-40834905">[13 more]</label></div><br/><div class="children"><div class="content">&gt; So you think you know what an “exception” means?<p>An exception is when code can no longer do any meaningful work.<p>This could be because of a programming bug, unexpected input, network issue, hardware issue, etc.  The cause doesn&#x27;t really matter.  What matters is that there is no way forward for the code.<p>And this is a <i>decision</i> that is made (or not made) by the programmer.  In this author&#x27;s Python example, he <i>chooses</i> accept the input as correct in the rest of the code.  If a valid URL is not provided then the code cannot continue because it is <i>assumed</i> correct.  One could easily change this code to have it continue; have it check for a invalid URL and reprint the prompt.  However, this would be a different program with different requirements.</div><br/><div id="40835410" class="c"><input type="checkbox" id="c-40835410" checked=""/><div class="controls bullet"><span class="by">kelnos</span><span>|</span><a href="#40834905">parent</a><span>|</span><a href="#40835113">next</a><span>|</span><label class="collapse" for="c-40835410">[-]</label><label class="expand" for="c-40835410">[5 more]</label></div><br/><div class="children"><div class="content">&gt; <i>An exception is when code can no longer do any meaningful work.</i><p>That is quite obviously not true, not even on the surface.  I have written tons of code where I&#x27;ve caught exceptions and then continued to do meaningful work.<p>The author&#x27;s Python example even contradicts your statement.  Sure, maybe they should validate the URL first.  But even if they did that, the server could be down, the user&#x27;s internet could be down, or there could be a typo in the URL that makes it unreachable, even if the URL itself is still valid.<p>And that doesn&#x27;t <i>have</i> to mean the code can no longer do meaningful work.  Maybe the exception handler could consult a local disk cache, and read the contents of the cache and display that instead.<p>I think exceptions <i>could</i> (and perhaps should) only be used when code can no longer do any meaningful work, but in practice that is very much not how real-world software works.<p>Ultimately, though, I prefer languages that don&#x27;t have exceptions.  Errors should be passed as return values, preferably using language features that make it impossible to forget to handle them.  If they truly get into a state where they can&#x27;t (or it&#x27;s unsafe to) do more meaningful work, then they should abort() and the program should be terminated.</div><br/><div id="40835677" class="c"><input type="checkbox" id="c-40835677" checked=""/><div class="controls bullet"><span class="by">tsimionescu</span><span>|</span><a href="#40834905">root</a><span>|</span><a href="#40835410">parent</a><span>|</span><a href="#40835420">next</a><span>|</span><label class="collapse" for="c-40835677">[-]</label><label class="expand" for="c-40835677">[1 more]</label></div><br/><div class="children"><div class="content">The code that <i>catches</i> an exception can continue, but the point was about the code that threw the exception. You <i>throw</i> an exception if you can&#x27;t continue the work you&#x27;re currently doing, for whatever reason. Of course, the rest of the program can very well continue: some other part will catch the exception you threw and decide what to do next. Ending the entire process is an extreme case, that should be reserved for only global problems - out of heap space errors, memory corruption, security violations, deadlock detection. <i>Anything</i> else should result in a more localized failure - killing a thread, a task, a request processor etc.<p>Also, there is little fundamentally different between throwing an exception and returning an error result. Everything I wrote above applies just as much to error results as it does to exceptions. Exceptions are just a language-level mechanism to help with a common error pattern: the code that detects an error condition is typically very far away from the code that can handle that error condition. When an error happens, the vast majority of the time, what happens next is that the error will be cascaded back through the call chain to a handler function of some kind. Along the way, some context will be added to the error, and some resources will be cleaned up. You can do this by hand, as you would in Go or Rust, or you can have the runtime do it for you, like in Java or C#. There are pros and cons to both (implicit vs explicit, how relevant is the context, etc), but fundamentally the program will take the exact same path.</div><br/></div></div><div id="40835420" class="c"><input type="checkbox" id="c-40835420" checked=""/><div class="controls bullet"><span class="by">wvenable</span><span>|</span><a href="#40834905">root</a><span>|</span><a href="#40835410">parent</a><span>|</span><a href="#40835677">prev</a><span>|</span><a href="#40835113">next</a><span>|</span><label class="collapse" for="c-40835420">[-]</label><label class="expand" for="c-40835420">[3 more]</label></div><br/><div class="children"><div class="content">&gt; But even if they did that, the server could be down, the user&#x27;s internet could be down, or there could be a typo in the URL that makes it unreachable, even if the URL itself is still valid.<p>That doesn&#x27;t contradict my statement.  In all those cases, the code cannot continue to do anymore meaningful work.  Handling the parsing error just handles that one situation.<p>&gt; I have written tons of code where I&#x27;ve caught exceptions and then continued to do meaningful work.<p>But the code that <i>threw the exception</i> could not continue to do meaningful work.  You can the catch the exception and <i>do something else</i> or maybe try the same operation again.  But either way, the original code could not and did not continue.  In general, you have to reach higher up in the call stack before you find a spot where you can something different or retry.  This is why exceptions are useful.</div><br/><div id="40835453" class="c"><input type="checkbox" id="c-40835453" checked=""/><div class="controls bullet"><span class="by">kelnos</span><span>|</span><a href="#40834905">root</a><span>|</span><a href="#40835420">parent</a><span>|</span><a href="#40835113">next</a><span>|</span><label class="collapse" for="c-40835453">[-]</label><label class="expand" for="c-40835453">[2 more]</label></div><br/><div class="children"><div class="content">&gt; <i>That doesn&#x27;t contradict my statement. In all those cases, the code cannot continue to do meaningful work. Just fixing the parsing error just takes one such condition out. I never claimed otherwise.</i><p>If you continued reading to the next paragraph, I presented an option for how the code <i>could</i> continue doing more work.<p>&gt; <i>But the code that threw the exception could not continue to do meaningful work. You can the catch the exception and do something else or maybe try the same operation again. But either way, the original code could not continue.</i><p>But that&#x27;s not what you said!  Or at least what you said was ambiguous.  &quot;The code can no longer do any meaningful work.&quot;  What code?  I interpreted that as &quot;all the following code&quot;, which feels like an entirely reasonable interpretation to me.  I feel like this just proves the author&#x27;s point -- it&#x27;s ambiguous!<p>Ok, so then what you really mean is: &quot;the code that threw the exception can no longer do any meaningful work&quot;.  Ok, sure, that&#x27;s true, but I don&#x27;t think that&#x27;s a useful definition.  I suppose perhaps it&#x27;s useful if you&#x27;re explaining it to a non-programmer who is unlikely to need any more information (arguably, you shouldn&#x27;t bring up exceptions to a non-programmer; just tell them &quot;the code encountered an error and couldn&#x27;t continue anymore&quot;).  But if you&#x27;re talking to a programmer, or, rather, teaching a new programmer, that definition is entirely insufficient.  Because from there, you need to talk about <i>why</i> the way to signal that inability is to use an exception.  You need to talk about other ways to signal that, and what the trade offs are.  You need to dig deeper.  At <i>best</i> your definition is only the very very very beginning of a longer conversation; it doesn&#x27;t give enough information on its own.</div><br/><div id="40835526" class="c"><input type="checkbox" id="c-40835526" checked=""/><div class="controls bullet"><span class="by">wvenable</span><span>|</span><a href="#40834905">root</a><span>|</span><a href="#40835453">parent</a><span>|</span><a href="#40835113">next</a><span>|</span><label class="collapse" for="c-40835526">[-]</label><label class="expand" for="c-40835526">[1 more]</label></div><br/><div class="children"><div class="content">&gt; What code? I interpreted that as &quot;all the following code&quot;, which feels like an entirely reasonable interpretation to me.<p>It is.  As long as you define &quot;all following code&quot; as the code that would execute from that point forward.  If you define &quot;following&quot; as simply the code in the rest of the file then doesn&#x27;t really make sense in this context I don&#x27;t why you would mean that.<p>Obviously the existence of catch statements allows the program to continue <i>from that point</i> but none of that code is following from where the error was (unless you call it again).<p>&gt; At best your definition is only the very very very beginning of a longer conversation; it doesn&#x27;t give enough information on its own.<p>If you want me to write an essay on this, I have actually already written one.  But I still think this definition is a good one.  It <i>is</i> simple but that&#x27;s what makes it useful.  The entire subject of error handling cannot be distilled into a single sentence definition and that is not what I intended to do.</div><br/></div></div></div></div></div></div></div></div><div id="40835113" class="c"><input type="checkbox" id="c-40835113" checked=""/><div class="controls bullet"><span class="by">hn_throwaway_99</span><span>|</span><a href="#40834905">parent</a><span>|</span><a href="#40835410">prev</a><span>|</span><a href="#40835763">next</a><span>|</span><label class="collapse" for="c-40835113">[-]</label><label class="expand" for="c-40835113">[7 more]</label></div><br/><div class="children"><div class="content">FWIW I didn&#x27;t particularly like the article, but I really liked your comment. I think &quot;an exception is when code can no longer do any meaningful work&quot; is a great definition, and also explains why flow control with exception handling works the way it does.<p>Obviously there are gray areas re what it is to mean &quot;code can no longer do any meaningful work&quot;, which is why there are decisions to be made about when to throw an exception or when to do something else. Usually, though, that &quot;something else&quot; is to return an error sentinel value, most often null or similar (e.g. &quot;undefined&quot;). And then I think the reason it&#x27;s sometimes a tough call is that since null&#x2F;undefined essentially means &quot;empty&quot;, the APIs that most often return null do so in cases where you&#x27;re asking for something and it&#x27;s not there, so returning &quot;empty&quot; feels reasonable.</div><br/><div id="40835620" class="c"><input type="checkbox" id="c-40835620" checked=""/><div class="controls bullet"><span class="by">lelanthran</span><span>|</span><a href="#40834905">root</a><span>|</span><a href="#40835113">parent</a><span>|</span><a href="#40835277">next</a><span>|</span><label class="collapse" for="c-40835620">[-]</label><label class="expand" for="c-40835620">[2 more]</label></div><br/><div class="children"><div class="content">&gt;  I think &quot;an exception is when code can no longer do any meaningful work&quot; is a great definition,<p>I&#x27;m not so sure: that definition glibly turns <i>every</i> error or failure into an exception.<p>Is &quot;Failure to Open File&quot; really an exceptional condition? This is a common condition that is basically a business rule[1].<p>Think about the &quot;saving a file&quot; action. There&#x27;s errors that you expect to happen, and have to provide code to handle (like, permission denied, path exists and is a directory, etc), and exceptional conditions that should never have arisen in a program which no sane person would try to handle (Out of memory, disk write errors, etc).<p>In both cases, no progress can be made and the callee has to return to the caller. But, in the error cases a well-written program will defend against the errors in a meaningful manner, whereas in the exceptional cases there is literally no point in putting in code to handle those cases.<p>Just some food for thought - I&#x27;m not particularly married to these definitions.<p>[1] Here are some hypothetical options for failing to open a file:<p>1. If the file doesn&#x27;t exist, prompt the user asking if it must be created, and proceed only if &quot;Yes&quot; is received, return to caller if &quot;No&quot; is received.<p>2. If the file is a directory, display a fatal error message and return to caller.<p>3. If permission is denied, display an ACL error message and return to caller.</div><br/><div id="40835665" class="c"><input type="checkbox" id="c-40835665" checked=""/><div class="controls bullet"><span class="by">wvenable</span><span>|</span><a href="#40834905">root</a><span>|</span><a href="#40835620">parent</a><span>|</span><a href="#40835277">next</a><span>|</span><label class="collapse" for="c-40835665">[-]</label><label class="expand" for="c-40835665">[1 more]</label></div><br/><div class="children"><div class="content">I have a desktop application with a single error handler at its event loop.  All that handler does is show the exception message in a messagebox and then the event loop resumes.  This application has a &quot;save file&quot; option.<p>In this application it doesn&#x27;t matter if the permission is denied, the path exists and is a directory, it was a network share that went away, or a disk write error, or out of memory, or a bug.  If I press &quot;Save&quot; and the exception occurs, I can read the message and correct the problem if possible, and then just press Save again.  It is incredibly robust and has effectively no error handling.<p>Of course, that doesn&#x27;t handle all the cases that you describe.  For example, for your case #1 I think that <i>isn&#x27;t</i> a case that no more meaningful work can be done and isn&#x27;t an error.  Instead, you&#x27;re just asking a question and then proceeding or not based on the answer.</div><br/></div></div></div></div><div id="40835277" class="c"><input type="checkbox" id="c-40835277" checked=""/><div class="controls bullet"><span class="by">wvenable</span><span>|</span><a href="#40834905">root</a><span>|</span><a href="#40835113">parent</a><span>|</span><a href="#40835620">prev</a><span>|</span><a href="#40835763">next</a><span>|</span><label class="collapse" for="c-40835277">[-]</label><label class="expand" for="c-40835277">[4 more]</label></div><br/><div class="children"><div class="content">In .NET, they have 2 sets of parsing functions, &quot;Parse&quot; and &quot;TryParse&quot; for parsing strings into other types.  The first throws an exception if it can&#x27;t parse and the second returns a Boolean indicating success or failure.<p>Which function of these you use as a programmer is dependent on the context and it indicates your expectation.  You might use &quot;Parse&quot; if you reading a well-defined file format where the string will always contain an integer and if it doesn&#x27;t then something has gone horribly wrong.  But you might use &quot;TryParse&quot; on user input where getting a non-integer is expected and you&#x27;ll have to something about that other than error out.</div><br/><div id="40835360" class="c"><input type="checkbox" id="c-40835360" checked=""/><div class="controls bullet"><span class="by">erik_seaberg</span><span>|</span><a href="#40834905">root</a><span>|</span><a href="#40835277">parent</a><span>|</span><a href="#40835763">next</a><span>|</span><label class="collapse" for="c-40835360">[-]</label><label class="expand" for="c-40835360">[3 more]</label></div><br/><div class="children"><div class="content">Was this a workaround for frequent exception handling being too expensive on some platform? I like how Java lets some exception instances be reused without stack frames, so performance issues are less likely to compromise API design.</div><br/><div id="40835673" class="c"><input type="checkbox" id="c-40835673" checked=""/><div class="controls bullet"><span class="by">jkrejcha</span><span>|</span><a href="#40834905">root</a><span>|</span><a href="#40835360">parent</a><span>|</span><a href="#40835401">next</a><span>|</span><label class="collapse" for="c-40835673">[-]</label><label class="expand" for="c-40835673">[1 more]</label></div><br/><div class="children"><div class="content">It, according to Eric Lippert, was at least partially to fix the problem of vexing exceptions[1] that plagued Int32.Parse. Parse was first and it was kind of an unfortunate design decision to make a almost expected result (parsing to fail) throw an exception.<p>It&#x27;s worth noting that nowadays Parse is just a wrapper around TryParse that throws (although this is an implementation detail)<p>[1]: <a href="https:&#x2F;&#x2F;ericlippert.com&#x2F;2008&#x2F;09&#x2F;10&#x2F;vexing-exceptions&#x2F;" rel="nofollow">https:&#x2F;&#x2F;ericlippert.com&#x2F;2008&#x2F;09&#x2F;10&#x2F;vexing-exceptions&#x2F;</a></div><br/></div></div><div id="40835401" class="c"><input type="checkbox" id="c-40835401" checked=""/><div class="controls bullet"><span class="by">wvenable</span><span>|</span><a href="#40834905">root</a><span>|</span><a href="#40835360">parent</a><span>|</span><a href="#40835673">prev</a><span>|</span><a href="#40835763">next</a><span>|</span><label class="collapse" for="c-40835401">[-]</label><label class="expand" for="c-40835401">[1 more]</label></div><br/><div class="children"><div class="content">Semantically I think it&#x27;s great. .NET has a bunch of these especially in LINQ.<p>A parsing failure isn&#x27;t necessary an error.  Sometimes it is, sometimes it&#x27;s not.  Even if exceptions weren&#x27;t expensive, catching exceptions on all parsing (as you would in Java) is just ugly and, I would argue, not semantically correct.  In my opinion, one should throw often but catch infrequently.  Most of my large applications have a very small number of catches.<p>It&#x27;s also perfectly reasonable to move this down a layer and have Parse return an Option type and choose to handle did-not-parse case or don&#x27;t and let that trigger an exception.</div><br/></div></div></div></div></div></div></div></div></div></div><div id="40835763" class="c"><input type="checkbox" id="c-40835763" checked=""/><div class="controls bullet"><span class="by">Kwpolska</span><span>|</span><a href="#40834905">prev</a><span>|</span><a href="#40835391">next</a><span>|</span><label class="collapse" for="c-40835763">[-]</label><label class="expand" for="c-40835763">[1 more]</label></div><br/><div class="children"><div class="content">&gt; Here’s the big lie: in Python, this snippet will almost always raise an Exception. That’s because most users, faced with an input prompt, will press Ctrl+C or enter a non-URL. And even if they enter a URL, it probably has a typo.<p>[citation needed]<p>&gt; In English, “exception” means, “something unusual.” But if this code snipet raises ValueError 60 per cent of the time and prints the contents of a web page 10 per cent of the time … well, surely we need a new term?<p>The purpose of this code snippet is to download a web page and show it to the user. The code considers situations that lead to it not fulfilling its purpose “exceptional”. If someone exits when they see a prompt for a URL, or type gibberish, why did they run a program to download a web page in the first place?</div><br/></div></div><div id="40835391" class="c"><input type="checkbox" id="c-40835391" checked=""/><div class="controls bullet"><span class="by">_ZeD_</span><span>|</span><a href="#40835763">prev</a><span>|</span><a href="#40835930">next</a><span>|</span><label class="collapse" for="c-40835391">[-]</label><label class="expand" for="c-40835391">[1 more]</label></div><br/><div class="children"><div class="content">&gt;&gt;&gt; When I shifted from Rust to Python, I tried to write matcher-style code. It wasn’t legible. I became irate.<p>so, in other words... rust developer tried to write rust in python, and it didn&#x27;t work.<p>wasn&#x27;t that the same story for each combination of two languages? I remember people criticize other people for &quot;always writing java in any language they work with&quot;...</div><br/></div></div><div id="40835930" class="c"><input type="checkbox" id="c-40835930" checked=""/><div class="controls bullet"><span class="by">xg15</span><span>|</span><a href="#40835391">prev</a><span>|</span><a href="#40835607">next</a><span>|</span><label class="collapse" for="c-40835930">[-]</label><label class="expand" for="c-40835930">[1 more]</label></div><br/><div class="children"><div class="content">You can have the same philosophical distinction over the term &quot;error&quot;. Is it really an error if your computer did not literally catch fire or otherwise took physical damage?<p>And yet this hasn&#x27;t stopped even nontechnical people from using the term without any significant communication problems.<p>I think a more interesting discussion might be what exactly the difference between &quot;errors&quot; and &quot;exceptions&quot; is in computing, though I&#x27;m not sure there is one in practice. (I <i>think</i> in theory, &quot;errors&quot; are supposed to be caused by factors external to your program, like hitting memory constraints, while &quot;exceptions&quot; are internal to your program? I&#x27;m not sure if this is universally true though or just a java-ism - and even in java it&#x27;s not consistently followed through everywhere).<p>So until this is decided, I think the most useful practical definition of exceptions is &quot;it&#x27;s a special language construct that makes it easier to manage errors&quot;.</div><br/></div></div><div id="40835607" class="c"><input type="checkbox" id="c-40835607" checked=""/><div class="controls bullet"><span class="by">lloeki</span><span>|</span><a href="#40835930">prev</a><span>|</span><a href="#40835305">next</a><span>|</span><label class="collapse" for="c-40835607">[-]</label><label class="expand" for="c-40835607">[1 more]</label></div><br/><div class="children"><div class="content">&gt; Python, C++, Ruby and Java all use exceptions. To them, “exception” means, “nifty piece of syntax kinda like goto.”<p>Not Ruby. In idiomatic Ruby^ exceptions are for exceptional cases, not code flow. &quot;nifty piece of syntax kinda like goto.&quot; is the much underused but very powerful catch+throw.<p>Or rather, it&#x27;s a stack unwinding mechanism: think baseball, the catcher expects a ball, with the ball being a message to be passed. Upon that you <i>could</i> implement a specialized and more costly case of stack unwinding which does what fail&#x2F;raise+rescue does, but catch+throw is more general than exceptions.<p>Take Python&#x27;s StopIteration as an example: there&#x27;s absolutely no need for this to be a full exception, instead it&#x27;s like `break`ing a loop but across method calls.<p>^ Nevermind Rails practices. Rails is not Ruby.</div><br/></div></div><div id="40835305" class="c"><input type="checkbox" id="c-40835305" checked=""/><div class="controls bullet"><span class="by">Mathnerd314</span><span>|</span><a href="#40835607">prev</a><span>|</span><a href="#40835296">next</a><span>|</span><label class="collapse" for="c-40835305">[-]</label><label class="expand" for="c-40835305">[1 more]</label></div><br/><div class="children"><div class="content">I will paraphrase the Wikipedia definition (itself from Flaviu Cristian and John B. Goodenough):<p>Each procedure has a precondition, a set of circumstances for which it will terminate &quot;normally&quot;. The set of &quot;normal&quot; circumstances is subjective and defined entirely by the programmer. An exception handling mechanism allows the procedure to &quot;raise an exception&quot; if this precondition is violated. The exception handling mechanism then specifies how the exception is handled.</div><br/></div></div><div id="40835296" class="c"><input type="checkbox" id="c-40835296" checked=""/><div class="controls bullet"><span class="by">winstonewert</span><span>|</span><a href="#40835305">prev</a><span>|</span><a href="#40835663">next</a><span>|</span><label class="collapse" for="c-40835296">[-]</label><label class="expand" for="c-40835296">[3 more]</label></div><br/><div class="children"><div class="content">My wish: people stop using the dumb line that exceptions are gotos.</div><br/><div id="40835327" class="c"><input type="checkbox" id="c-40835327" checked=""/><div class="controls bullet"><span class="by">wvenable</span><span>|</span><a href="#40835296">parent</a><span>|</span><a href="#40835663">next</a><span>|</span><label class="collapse" for="c-40835327">[-]</label><label class="expand" for="c-40835327">[2 more]</label></div><br/><div class="children"><div class="content">The pattern for returning errors is something like the following pseudocode:<p><pre><code>    main() 
    {
         (result or error) = functionA();
         if (error) die(&quot;There was an error: &quot; + error);
         print result
    }

    functionA() 
    {
       (result1 or error1) = functionB()
       if (error1) return error1;
       (result2 or error2) = functionC(result1);
       if (error2) return error2;
       (result3 or error3) = functionD(result2);
       return result3;
    }

     functionB()
     {
         (result or error) = readFileAndGetResult()
         if (error) return error;
         return result;
     }
</code></pre>
Exceptions simply automate the error returns; the same code with exceptions is merely:<p><pre><code>    main()
    {
        try {
            result = functionA();
            print result;
         } catch (error) {
             die(&quot;There was an error: &quot; + error);
         }
     }

     functionA()
     {
         result1 = functionB()
         result2 = functionC(result1);
         result3 = functionD(result2);
         return result3;    
     }

     functionB()
     {
         return readFileAndGetResult();
     }
</code></pre>
The most naive implementation of exceptions would do exactly what the top code does.  Check for an error result from the function and return it to it&#x27;s caller until it finds a catch block.<p>The advantage of exceptions is that you don&#x27;t get the bug that exists in functionA where it fails to propagate the error from functionD.  It also makes the logic of the code much more readable and writable because you don&#x27;t need to put an error guard around the call to every single function.<p>But what is telling is that, in both examples, the error is handled in <i>exactly</i> the same place in the code.  How can one reasonably say the second case is like a goto but the first isn&#x27;t?</div><br/></div></div></div></div><div id="40835663" class="c"><input type="checkbox" id="c-40835663" checked=""/><div class="controls bullet"><span class="by">kaba0</span><span>|</span><a href="#40835296">prev</a><span>|</span><a href="#40835682">next</a><span>|</span><label class="collapse" for="c-40835663">[-]</label><label class="expand" for="c-40835663">[2 more]</label></div><br/><div class="children"><div class="content">I think this topic has been discussed countless times, and this post fails to add anything worthy to it.<p>My two cents: use result types, or their analog in your language for expected error conditions, and use exceptions for exceptional situations. Example for the former is parsing, which has an expected failure mode - something couldn’t be parsed as intended. Example for the latter is a network issue during an API call&#x2F;writing a file&#x2F;etc. Nonetheless, they are a homologous structure to a Result type in more FP languages, so it’s purely which is a better fit for the given circumstance.<p>I would also like to defend exceptions, as they are a very good construction with sane and safe defaults. They let people concentrate on the actual business logic, not sprinkling little bit of business and error handling logic all over the place; they do the sane thing by default: auto-unwrap the correct result, auto-bubble up if not handled at a given scope; and often unappreciated: they allow as fine- or course-grained error handling as needed, while it is only possible with some less than readable Monad magic with Result types. And most importantly, they never let an erroneous condition silently go unnoticed, you have to deliberately ignore them for that. With errno checking (c, go), and a bit less so with result types you can fall victim to that.</div><br/><div id="40835928" class="c"><input type="checkbox" id="c-40835928" checked=""/><div class="controls bullet"><span class="by">MaxBarraclough</span><span>|</span><a href="#40835663">parent</a><span>|</span><a href="#40835682">next</a><span>|</span><label class="collapse" for="c-40835928">[-]</label><label class="expand" for="c-40835928">[1 more]</label></div><br/><div class="children"><div class="content">&gt; They let people concentrate on the actual business logic, not sprinkling little bit of business and error handling logic all over the place<p>This can lead to trouble if the programmer doesn&#x27;t keep in mind that various (but perhaps not all) operations can throw. In particular, proper behaviour in the exceptional case may require a more specific order of operations than is needed for proper behaviour in the normal case.<p>Raymond Chen did a great blog post on this in 2005, <i>Cleaner, more elegant, and harder to recognize</i>. [0][1]<p>[0] <a href="https:&#x2F;&#x2F;devblogs.microsoft.com&#x2F;oldnewthing&#x2F;20050114-00&#x2F;?p=36693" rel="nofollow">https:&#x2F;&#x2F;devblogs.microsoft.com&#x2F;oldnewthing&#x2F;20050114-00&#x2F;?p=36...</a><p>[1] <a href="https:&#x2F;&#x2F;news.ycombinator.com&#x2F;item?id=406391">https:&#x2F;&#x2F;news.ycombinator.com&#x2F;item?id=406391</a></div><br/></div></div></div></div><div id="40835682" class="c"><input type="checkbox" id="c-40835682" checked=""/><div class="controls bullet"><span class="by">TheTrueScotsman</span><span>|</span><a href="#40835663">prev</a><span>|</span><label class="collapse" for="c-40835682">[-]</label><label class="expand" for="c-40835682">[1 more]</label></div><br/><div class="children"><div class="content">Exceptions are the hallmark of the midwit, no true developer would use them.</div><br/></div></div></div></div></div></div></div></body></html>